sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -698,28 +1147,147 @@ pi_result piMemGetInfo(pi_mem, cl_mem_info, size_t, void *, size_t *) {
+    int result =
+        Mem->Context->CmDevicePtr->DestroySurface(pi_img->CmSurfacePtr);
+    if (result != cm_support::CM_SUCCESS) {
+      return PI_INVALID_MEM_OBJECT;
+    }
+  } else {
+    return PI_INVALID_MEM_OBJECT;
+  }
+
   return PI_SUCCESS;
 }
 
+cm_support::CM_SURFACE_FORMAT
+ConvertPiImageFormatToCmFormat(const pi_image_format *piFormat) {
+  using ULongPair = std::pair<unsigned long, unsigned long>;
+  using FmtMap = std::map<ULongPair, cm_support::CM_SURFACE_FORMAT>;
+  static const FmtMap pi2cm = {
+      {{PI_IMAGE_CHANNEL_TYPE_UNORM_INT8, PI_IMAGE_CHANNEL_ORDER_RGBA},
+       cm_support::CM_SURFACE_FORMAT_A8R8G8B8},
+
+      {{PI_IMAGE_CHANNEL_TYPE_UNORM_INT8, PI_IMAGE_CHANNEL_ORDER_ARGB},
+       cm_support::CM_SURFACE_FORMAT_A8R8G8B8},
+
+      {{PI_IMAGE_CHANNEL_TYPE_UNSIGNED_INT8, PI_IMAGE_CHANNEL_ORDER_RGBA},
+       cm_support::CM_SURFACE_FORMAT_A8R8G8B8},
+
+      {{PI_IMAGE_CHANNEL_TYPE_UNSIGNED_INT32, PI_IMAGE_CHANNEL_ORDER_RGBA},
+       cm_support::CM_SURFACE_FORMAT_R32G32B32A32F},
+  };
+  auto result = pi2cm.find(
+      {piFormat->image_channel_data_type, piFormat->image_channel_order});
+  if (result != pi2cm.end()) {
+    return result->second;
+  }
+  DIE_NO_IMPLEMENTATION;
+  return cm_support::CM_SURFACE_FORMAT_A8R8G8B8;
+}
+
 pi_result piMemImageCreate(pi_context Context, pi_mem_flags Flags,
                            const pi_image_format *ImageFormat,
                            const pi_image_desc *ImageDesc, void *HostPtr,
                            pi_mem *RetImage) {
-  (void)Context;
-  (void)Flags;
-  (void)ImageFormat;
-  (void)ImageDesc;
-  (void)HostPtr;
-  (void)RetImage;
-  DIE_NO_IMPLEMENTATION;
+  if (ImageFormat == nullptr || ImageDesc == nullptr)
+    return PI_INVALID_VALUE;
+
+  switch (ImageDesc->image_type) {
+  case PI_MEM_TYPE_IMAGE2D:
+    break;
+  default:
+    return PI_INVALID_MEM_OBJECT;
+  }
+
+  auto bytesPerPixel = 4;
+  switch (ImageFormat->image_channel_data_type) {
+  case PI_IMAGE_CHANNEL_TYPE_UNSIGNED_INT32:
+    bytesPerPixel = 16;
+    break;
+  case PI_IMAGE_CHANNEL_TYPE_UNSIGNED_INT8:
+  case PI_IMAGE_CHANNEL_TYPE_UNORM_INT8:
+    bytesPerPixel = 4;
+    break;
+  default:
+    return PI_INVALID_VALUE;
+  }
+
+  cm_support::CmSurface2D *CmSurface = nullptr;
+  cm_support::SurfaceIndex *CmIndex;
+
+  int status = Context->CmDevicePtr->CreateSurface2D(
+      static_cast<unsigned int>(ImageDesc->image_width),
+      static_cast<unsigned int>(ImageDesc->image_height),
+      ConvertPiImageFormatToCmFormat(ImageFormat), CmSurface);
+
+  if (status != cm_support::CM_SUCCESS) {
+    return PI_OUT_OF_HOST_MEMORY;
+  }
+
+  status = CmSurface->GetIndex(CmIndex);
+
+  if ((Flags & PI_MEM_FLAGS_HOST_PTR_USE) != 0 ||
+      (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) != 0) {
+
+    if (HostPtr != nullptr) {
+
+      status = CmSurface->WriteSurface(
+          reinterpret_cast<const unsigned char *>(HostPtr), nullptr,
+          static_cast<unsigned int>(ImageDesc->image_width *
```

Comment:
[Q]
Is it possible for CM to reuse HostPtr or a separate allocation + copy is absolutely required here?

Reply:
I think reusing is possible.

Reply:
So, reusing sounds like a good optimization here.

Reply:
I will apply optimization after the ESIMD_CPU support becomes fully functional.